### PR TITLE
This PR spells out the rollout states

### PR DESCRIPTION
**Pull request by ryanzhang-oss**

### CODE REVIEW

File: pkg/controller/common/rollout/rollout_state.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+)
+
+const invalidRollingStateTransition = "the rollout state transition from `%s` state  with `%s` is invalid"
+
+const invalidBatchRollingStateTransition = "the batch rolling state transition from `%s` state  with `%s` is invalid"
+
+// StateTransition is the center place to do rollout state transition
+// it returns an error if the transition is invalid
+// it changes the coming rollout state if it's valid
+// nolint: gocyclo
```
Comment (wonderflow) [inline pkg/controller/common/rollout/rollout_state.go:66]: it's better to split into small functions

File: pkg/controller/common/rollout/rollout_state.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+func StateTransition(rolloutStatus *v1alpha1.RolloutStatus, event rolloutEvent) error {
```
Comment (wonderflow) [inline pkg/controller/common/rollout/rollout_state.go:67]:
```suggestion
func StateMachine(rolloutStatus *v1alpha1.RolloutStatus, event rolloutEvent) error {
```

What about this name?

**Review by wonderflow — APPROVED** [review]

File: pkg/controller/common/rollout/rollout_state.go (hunk 2 above)
Comment (ryanzhang-oss) [reply pkg/controller/common/rollout/rollout_state.go:67]: either way is fine.

File: pkg/controller/common/rollout/workloads/factory.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+		rolloutSpec:    f.rolloutSpec,
+		targetWorkload: f.targetWorkload,
+	}
+
+	switch kind.Kind {
+	case "CloneSet":
+		return cloneSetCtrl
+
+	default:
+		return cloneSetCtrl
```
Comment (zzxwill) [inline pkg/controller/common/rollout/workloads/factory.go:45]: These two cases return a same controller. Is it by design?

Comment (ryanzhang-oss) [reply pkg/controller/common/rollout/workloads/factory.go:45]: yes and no, this is just a framework.

File: pkg/controller/common/rollout/rollout_state.go (hunk 1 above)
Comment (ryanzhang-oss) [reply pkg/controller/common/rollout/rollout_state.go:66]: done

**Pull request merged**